dotnet/src/Connectors/Connectors.Memory.Kusto/KustoMemoryStore.cs

```diff
@@ -0,0 +1,393 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Data;
+using System.Globalization;
+using Kusto.Data;
+using Kusto.Data.Common;
+using Kusto.Data.Net.Client;
+using Microsoft.SemanticKernel.AI.Embeddings;
+using Microsoft.SemanticKernel.Memory;
+using Newtonsoft.Json;
+
+namespace Microsoft.SemanticKernel.Connectors.Memory.Kusto;
+
+internal static class KustoSerializer
+{
+    public static string SerializeEmbedding(Embedding<float> embedding)
+    {
+        return JsonConvert.SerializeObject(embedding.Vector);
+    }
+    public static Embedding<float> DeserializeEmbedding(string embedding)
+    {
+        if (string.IsNullOrEmpty(embedding))
+        {
+            return default;
+        }
+
+        float[] floatArray = JsonConvert.DeserializeObject<float[]>(embedding);
+        return new Embedding<float>(floatArray);
+    }
+    public static string SerializeMetadata(MemoryRecordMetadata metadata)
+    {
+        if (metadata == null)
+        {
+            return string.Empty;
+        }
+
+        return JsonConvert.SerializeObject(metadata).Replace("\"", "\"\"");
+    }
+    public static MemoryRecordMetadata? DeserializeMetadata(string metadata)
+    {
+        if (string.IsNullOrEmpty(metadata))
+        {
+            return default;
+        }
+
+        return JsonConvert.DeserializeObject<MemoryRecordMetadata>(metadata);
+    }
+}
+
+internal static class KustoExtensions
+{
+    public static async ValueTask<T?> FirstOrDefaultAsync<T>(this IAsyncEnumerable<T> source, CancellationToken cancellationToken = default)
+    {
+        await foreach (var item in source.WithCancellation(cancellationToken).ConfigureAwait(false))
+        {
+            return item;
+        }
+
+        return default;
+    }
+}
+
+internal class KustoMemoryRecord
+{
+    public string Key { get; set; }
+    public MemoryRecordMetadata Metadata { get; set; }
+    public Embedding<float> Embedding { get; set; }
+    public DateTime? Timestamp { get; set; }
+
+    public KustoMemoryRecord(MemoryRecord record) : this(record.Key, record.Metadata, record.Embedding, record.Timestamp?.UtcDateTime) { }
+
+    public KustoMemoryRecord(string key, MemoryRecordMetadata metadata, Embedding<float> embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = metadata;
+        this.Embedding = embedding;
+        this.Timestamp = timestamp;
+    }
+
+    public KustoMemoryRecord(string key, string metadata, string embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = KustoSerializer.DeserializeMetadata(metadata);
+        this.Embedding = KustoSerializer.DeserializeEmbedding(embedding);
+        this.Timestamp = timestamp;
+    }
+
+    public MemoryRecord ToMemoryRecord()
+    {
+        return new MemoryRecord((MemoryRecordMetadata)this.Metadata, (Embedding<float>)this.Embedding, this.Key, this.Timestamp);
+    }
+
+    public string ToCsvString()
```

Comment:
>ToCsvString [](http://example.com/codeflow?start=18&length=11)

We already have a CsvWriter functionality, don't recreate it (is this handling escaping characters for example?)

Reply:
changed per your recommendation